src/context.rs

```diff
@@ -105,10 +106,83 @@ fn test_should_add_a_service() {
         }
     }
 
-    let service = ServiceStub;
-    let mut foo = Context::new(false, Some("localhost".to_owned()), None, None);
+    mod add_service {
+        use super::super::*;
 
-    assert_eq!(foo.services.is_empty(), true);
-    foo.add_service(Box::new(service));
-    assert_eq!(foo.services.is_empty(), false);
+        #[test]
+        fn test_should_increase_number_of_services() {
+            let service = super::ServiceStub;
+            let context = Context::shared(false, Some("localhost".to_owned()), None, None);
+            let mut locked_context = context.lock().unwrap();
```

Comment:
It turns out one of the links given by @brson suits our current needs. I used [stainless](https://github.com/reem/stainless) which enables `before_each` macros and nesting.

Thanks @brson!